### ISSUE TITLE
CCAPS - No Slip + Grav issue

### ISSUE THREAD

**Issue by JOThurgood** [JOThurgood/SimpleCFD#101]
Description

- Set the RTI problem up with no density interface (e.g. rho = 1 throughout)
- Will see divergence + edge-mode v velocity emanate from top boundary (boundary opposite to grav drn)
- from printouts, correction to v is not correctly implemented near top boundary (at iy = ny-1 and iy = ny, but with the cells at iy=ny being a much larger error)

Cause

- the smaller errors at iy = ny-1 are a consequence of the BC choice of extrapolation of pressure through the boundary. Different BC choices / higher order interpolation / not using centred differences for evaluation of pressure gradient near the edges might improve this (I suspect not using CD but "inward difference" near the edge is the way to go so gradp is never determined based on boundary?)

- the larger error at iy = ny is mainly due to rho(ix,iy=ny) not remaining =1 at the boundary (although the choice of pressure bc also effects it somewhat)
- this is an unexpected result from `advect_dens` 

- Tracing it back, it emanates from the evaluation of the right (full/transversal) rho state)

```Fortran
705         rhoyr(ix,iy) = rhohyr(ix,iy) &
706           & - 0.5_num * dt * rho(ix,iy+1) * (macv(ix,iy+1) - macv(ix,iy)) / dy &
707           & - 0.5_num * dt / dx * ( rhohx(ix,iy+1)*macu(ix,iy+1) - &
708                                     & rhohx(ix-1,iy+1)*macu(ix-1,iy+1) )  
```

- Specifically the term at 706. 

- This is because you "expected"  `macv(ix,iy+1) - macv(ix,iy)`  to evaluate as 0 but no slip means that macv(iy+1) = - macv(iy), so you get a non zero term

- **So the boundary as is doesn't really make sense with the gravity?** Could either implement a zero-gradient / inoutflow boundary at the top, or implement (after this step, gradp should correctly compensate  the gravity and noslip on internal flows makes more sense?)

**Comment by JOThurgood** [comment]
- If you set ZG on rho and v, you will avoid the issue of rho dropping below at top
- Of course, you can't set ZG as the bc on phi, otherwise you'll be messing up the correction (by a factor of half).

**Comment by JOThurgood** [comment]
There is also a subtlety in that (I have the impression from literature) that often ZG on phi is a good / the best choice to pass to the multigrid solver. This could mean that you end up computing a phi around the edges that is polluted by the boundary somewhat (even if the interior L2 is good?) and then you'll not match grad p ???

**Comment by JOThurgood** [comment]
> Q - Why is there not an analogous issue at the bottom boundary ? There, you do get grad_p = rho g

Maybe this hints that the MG is doing fine (despite using ZG on mg phi bc -  which seems physically wrong - it could still converge to having the gradient in the domain ). Is it because only the top boundary exhibits the rho drop at the top cell, which in turn feeds through into MG via the 1/eta term, but the bottom remains rho = 1. 

If in step 5 you pass eta = 1/1 instead of 1/rho to the mg, you get a much better gradient for phi at the iy = ny cells

**Comment by JOThurgood** [comment]
So when ymin = no_slip and ymax= zero gradient you get basically a divergence free flow except at the bottom boundary where, due to v = 0 on it, but v = - whatever nearby, you get a whopping divergence. This was there previously, but matched by an equal and opposite divergence at the top.

The MG converges to something bad with a huge L2 to try and eliminate this single source of divergence. It wont give you a phi-profile s.t. grad p = rho g

So ZG fixes the problem with rho "dripping" from the top, and so opens the door to a better correction there, but opens up this can of worms.

I wonder if, in the non-atmosphere subtracted codes, whether it would be worthwhile requiring the user to give grad p as part of the initial condition, in lieu of bootstrapping? Or, perhaps don't advect mass during the initial step ?

**Comment by JOThurgood** [comment]
> when ymin = no_slip and ymax= zero gradient ... The MG converges to something bad with a huge L2 to try and eliminate this single source of divergence. It wont give you a phi-profile s.t. grad p = rho g

If, for the projection only, you set phi = 0 on the top boundary you'll get a linear phi profile out of MG.  Then, if you also set phi = 0 in the hydro bcs, should get a reasonable correction at all cells, provided rho =1 remains everywhere (which ZG on velocity and rho otherwise should do?)

You still get the divergence now only at the bottom, but the MG manages to come up with a phi that has a linear gradient of -1 (of grad rho)

**Comment by JOThurgood** [comment]
Created a boundary type "outflow", which is currently only specified for ymax.

For the described test problem it correctly prescribes gradp to balance the atmosphere and then appears to remains stable. (tested for nx=16, ny=4nx, for t_end = 100 and div <  1e-8 by end - slowly growing vertical stripes - for higher resolutions nx=64, 128 magnitude of div is larger (but still small really) and everything is stable).

(hydro) Velocity

```Fortran 
    ! outflow - zero gradient on v and rho, fixed phi

    if (bc_ymax == outflow) then

      if (present(arr_cc)) then
        arr_cc(:,ny+1) = arr_cc(:,ny)
        arr_cc(:,ny+2) = arr_cc(:,ny-1)
      endif

      if (present(arr_xface)) then
        arr_xface(:,ny+1) = arr_xface(:,ny)
        arr_xface(:,ny+2) = arr_xface(:,ny-1)
      endif

      if (present(arr_yface)) then
        arr_yface(:,ny+1) = arr_yface(:,ny-1)
        arr_yface(:,ny+2) = arr_yface(:,ny-2)
      endif

    endif
```

(hydro) Density

```Fortran
    if (bc_ymax == outflow) then

      if (present(arr_cc)) then
        arr_cc(:,ny+1) = arr_cc(:,ny)
        arr_cc(:,ny+2) = arr_cc(:,ny-1)
      endif

      if (present(arr_xface)) then
        arr_xface(:,ny+1) = arr_xface(:,ny)
        arr_xface(:,ny+2) = arr_xface(:,ny-1)
      endif

      if (present(arr_yface)) then
        arr_yface(:,ny+1) = arr_yface(:,ny-1)
        arr_yface(:,ny+2) = arr_yface(:,ny-2)
      endif

    endif
```


hydro Phi
```Fortran
    if (bc_ymax == outflow) then
      phi(:,ny+1) = phi(:,ny) + (phi(:,ny) - phi(:,ny-1))
      phi(:,ny+2) = phi(:,ny) + (phi(:,ny) - phi(:,ny-1))*2.0_num
    endif
```

multigrid phi and eta

```Fortran
    if (bc_ymax == outflow) then
      mg_bc_ymax = 'fixed'
      mg_etabc_ymax = 'zero_gradient'
    endif
```

Might be that there is a slight discrepancy between fixed eta in mg and then the extrapolation used for gradp calc in rest of code

**Comment by JOThurgood** [comment]
So in summary, this is basically because

- you want to enforce HSE in gravitational setups.

- Except in the background-subtracted model, this can only realistically be achieved by extrapolation. 

- but usually at a solid wall rho and pressure would be chosen as zero gradient

I've refactored the boundaries and created `no_slip` `slip` `no_slip_hse` and `slip_hse` as distinct entities.

As for outflow, this is tricky. Usually on an outflow you'd want to enforce tangential pressure is zero (no shear on boundary), but this is somewhat at odds with achieving hse via linear extrap.

I've created `outflow_hse`, which prioritises hse over no shear (but treats velocity and density as zero grad).  The idea is `outflow` could be implemented for non gravitational cases where the zero shear is respected. Additionally, for the atmosphere versions, this should be different.

Pushed to dev for CCAPS but not into the flavours, will be soon.
